### PR TITLE
Leaves - Julia K

### PR DESCRIPTION
**Pull request by Kalakalot**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a top ten method for the Work model that takes in a work category and returns the ten works in that category with the most votes, sorted descending (works with the most votes first). 
Describe how you approached testing that model method. What edge cases did you come up with? | I started with nominal positive cases like "it only adds works with one or more votes and works are in the expected order" and added a nominal negative case with "it doesn't add works with zero votes." The edge case is if there are no works in the given category with any votes.
What are `session` and `flash`? What is the difference between them? | Session is a hash-like object used to keep track of data throughout a user's session (which normally ends when they close their browser). Flash in Rails is a special type of hash used to send messages from a controller to a view. A flash message persists from one request to another, i.e., if you require a user to be logged in to vote, the user will see the "you must be logged in to do that" message after the view redirects them.
What was one thing that you gained more clarity on through this assignment? | I got a lot more comfortable inspecting an existing website for CSS and adding embedded Ruby to views. Creating the view pages was by far the easiest part of this assignment for me. 
What is the Heroku URL of your deployed application? | https://media-ranker-jak.herokuapp.com
Do you have any recommendations on how we could improve this project for the next cohort? |More classroom time with Bootstrap could be helpful, as could some review of the Block Element Modifier CSS naming convention (names like "btn btn--primary" made a lot more sense after I did some reading about BEM but it sounded like a lot of folks were confused by the names in inspected CSS).

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  
+  resources :works 
+  
+  get "/login", to: "users#login_form", as: "login"
+  post "/login", to: "users#login"
+  post "/logout", to: "users#logout", as: "logout"
+  get "/users/:id", to: "users#show", as: "user"
+  get "/users", to: "users#index", as: "users"
+
+  post "/works/:id/upvote", to: "votes#create", as: "upvote"
```
Comment (dHelmgren) [inline config/routes.rb:15]: Love the restraint here. Just the ones you need, exactly as you need em.

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        <% Work.votes_descending("album").each do |work| %>
+          <tr>
+            <td><%= Vote.where(work_id: work.id).count %></td>
+            <td><a href="/works/<%= work.id %>"> <%= work.title %>
+            </a></td>
+            <td><%= work.creator %></td>
+            <td><%= work.publication %></td>
+            <td><a class="btn btn-primary" rel="nofollow" data-method="post" href="/works/<%= work.id %>/upvote">Upvote</a></td>
+             <%end%>
+          </tr>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:25]: This section has a lot of repeated code with the other two categories. Could you use a view partial to dry this up?

File: test/fixtures/votes.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+#   user_id: 3
+#   work_id: 3
+
+# vote4:
+#   user_id: 4
+#   work_id: 4
+
+# vote5:
+#   user_id: 5
+#   work_id: 5
```
Comment (dHelmgren) [inline test/fixtures/votes.yml:19]: Why are these all commented out?

File: test/models/work_test.rb [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    it "must have a title" do
+      movie.title = nil
+      
+      expect(movie.valid?).must_equal false
+      expect(movie.errors.messages).must_include :title
+      expect(movie.errors.messages[:title]).must_equal ["can't be blank"]
+    end
+    
+    it "title must be unique For category" do
+      duplicate_title = Work.new(category: "movie", title: "Soup Dumplings", creator: "No one", publication: 1999, description: " ")
```
Comment (dHelmgren) [inline test/models/work_test.rb:47]: I would do this with a fixture, as that's a better proof.
```suggestion
      duplicate_title = Work.new(category: "movie", title: Work.first, creator: "No one", publication: 1999, description: " ")
```

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | no
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | no flash message!
Can go to "View all media" page and see three lists of works, sorted by vote | yes
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | no
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes 
Routes file uses `resources` for works | ues
Routes file shows intention in limiting routes for voting, log-in functionality, and users |yes 
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | no, see comments.
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ye
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | no, missing
Some controller, likely the `WorksController`, has a controller filter for finding a work | no, missing
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | no, see comments
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Good job overall! There are a few areas that you're missing right now, specifically using controller filters and understanding how to use view partials beyond forms. That said, the rest of this project is very strong!
